addons/website_sale/models/product_feed.py

```diff
@@ -0,0 +1,381 @@
+             " If not set, the default website pricelist will be used.\n"
+             "Note that the pricelist must be selectable on the website.",
+        domain="[('website_id', 'in', (False, website_id)), ('selectable', '=', True)]",
+    )
+    lang_id = fields.Many2one(
+        'res.lang',
+        string="Language",
+        help="Select the language to translate product names, descriptions,"
+             " and other text in the feed.",
+        required=True,
+        domain="[('id', 'in', website_lang_ids)]",
+    )
+    website_lang_ids = fields.Many2many(related='website_id.language_ids')
+    product_category_ids = fields.Many2many('product.public.category', string="Categories")
+    format = fields.Selection(
+        selection=[
+            ('gmc', "Google Merchant Center"),
+        ],
+        default='gmc',
+        required=True,
+    )
+    access_token = fields.Char(
+        default=lambda _: uuid.uuid4().hex,
+        required=True,
+        copy=False,
+        readonly=True,
+    )
+    url = fields.Char(compute='_compute_url')
+
+    # Caching mechanism (technical fields)
+    feed_cache = fields.Binary(compute='_compute_feed_cache', store=True, readonly=True)
+    cache_expiry = fields.Datetime(default=fields.Datetime.now, required=True, readonly=True)
+
+    @api.depends('format')
+    def _compute_url(self):
+        """Compute the full feed url."""
+        for feed in self:
+            match feed.format:
+                case 'gmc':
+                    path = '/gmc.xml'
+                case _:
+                    raise NotImplementedError()
+
+            feed.url = urls.urljoin(
+                feed.website_id.get_base_url(),
+                f'{path}?feed_id={feed.id}&access_token={feed.access_token}',
+            )
+
+    @api.depends(
+        'website_id', 'pricelist_id', 'lang_id', 'product_category_ids'
+    )
+    def _compute_feed_cache(self):
+        """Invalidate cache on feed parameter changes."""
+        self.action_invalidate_cache()
+
+    @api.constrains('product_category_ids', 'website_id')
+    def _check_product_limit(self):
+        """Add a soft limit on the number of products a feed can contain. A strong limit of 6000 is
+        applied during the feed rendering phase."""
+        for feed in self:
+            product_count = feed.env['product.product'].search_count(
+                feed._get_feed_product_domain(), limit=const.PRODUCT_FEED_SOFT_LIMIT + 1
+            )
+            if product_count > const.PRODUCT_FEED_SOFT_LIMIT:
+                raise ValidationError(feed.env._(
+                    "A single feed cannot contain more than %(limit)s products."
+                    " Please separate products with Categories.",
+                    limit=f"{const.PRODUCT_FEED_SOFT_LIMIT:,}",  # Format to 5,000
+                ))
+
+    def action_invalidate_cache(self):
+        self.feed_cache = False
```

Comment:
Don't drop feed cache, only update the cache expiry to earlier than now.
(would potentially allow cron management in the future)